feat(material-experimental): add test harness for form-field

devversion opened the pull request.

devversion commented on src/material-experimental/form-field/testing/form-field-harness.ts:
```diff
@@ -0,0 +1,229 @@
+  /** Gets the harness of the control that is bound to the form-field. */
+  async getControl(): Promise<FormFieldControlHarness|null> {
+    const environment: HarnessEnvironment<any> = (this as any).locatorFactory;
+    // TODO(reviewer): not sure what we should do here. We need to construct a combined
```
I need feedback on this. This is just a hack for now do demonstrate on what I have in mind. Any other ideas/or should we expose these things in the locator factory?

devversion replied:
Adding what I wrote on Slack. This would be my proposal:
```ts
private _controlSelector = this.locatorForAll('input,select');

// ----

const elements = await this._controlSelector();

if (!elements.length) {
  return null;
}
...
this.environment.createHarness(elements[0], HarnessType);
```

Basically something like that. It does not introduce too much changes in the harness infrastructure and the mental model remains quite similar IMO.

Two things are special here:
* Exposing the environment to the harnesses, so that they can create harnesses out of elements.
* And the other interesting thing is that createHarness creates the harness with an actual TestElement. It can do that because the corresponding environment knows how it can access the native element/ or it can just re-use the passed element.

jelbourn replied:
Wouldn't it be much simpler to have mat-form-field apply a marker CSS class to its control and look for that?

devversion replied:
Yeah that would be another possibility. Though it feels slightly wrong to add host bindings (which will be checked by Angular and affect micro-perf) in favor of a test harness.

I have the feeling that this is a pattern that people will hit more often (i.e. creating a harness of an element), so I'd vote for adding something like that anyway.

jelbourn replied:
It wouldn't be a binding, just a static class we add once here:
https://github.com/angular/components/blob/94c3fed9da74f3560467a4d702c240120a4c4c7f/src/material/form-field/form-field.ts#L300-L299

devversion replied:
Should we still keep track (somewhere else) of the use-case of creating a harness out of an element? cc. @mmalerba

mmalerba replied:
@devversion yeah why don't you file an issue to consider adding the feature and we'll see if anyone chimes in asking for it. It feels like a pretty niche case to me, so we may be able to do without it.

devversion replied:
Sounds good. will do.

jelbourn commented on src/material-experimental/form-field/testing/form-field-harness.ts:
```diff
@@ -0,0 +1,231 @@
+      return 'fill';
+    } else if (isOutline) {
+      return 'outline';
+    }
```
What do you think of condensing this to
```ts
const hostClass = await (await this.host()).getAttribute('class');
const appearanceMatch = hostClass.match(/mat-form-field-appearance-(\w+)/);
if (appearanceMatch) {
  return appearanceMatch[1];
}

throw Error('Could not determine appearance of form-field.');
```

devversion replied:
I like that it is way shorter and clean. Though it's not 100% accurate as it could also match accidental cases where the class looks like `mat-form-field-appearance-custom`.

We could fix that by adding the valid options to the capture group I guess.

mmalerba replied:
I think this regex should get just the cases we want:
`/mat-form-field-appearance-(legacy|standard|fill|outline)(?:$|\s)/`

devversion replied:
:+1: and good point with matching either end or a space!

devversion replied:
Done.

jelbourn commented on src/material-experimental/form-field/testing/form-field-harness.ts:
```diff
@@ -0,0 +1,231 @@
+    } else if (isWarn) {
+      return 'warn';
+    }
+    return 'primary';
```
A bit more concise?
```ts
const hostClass = await (await this.host()).getAttribute('class');
for (const palette of ['accent', 'warn']) {
  if (hostClass.includes(palette)) {
    return palette;
  }
}

return 'primary';
```

devversion replied:
Sounds like a good idea.

devversion replied:
I'll actually keep it like that as otherwise we'd need to handle more things (i.e. spaces to separate classes in the  `class` attribute). The current implementation is quite reasonable but just looks weird due to the Promise all for concurrency.

mmalerba replied:
You could maybe use `getProperty('classList')` for this, now that we have separate methods for attributes and properties

jelbourn replied:
I don't follow what you mean by handing more things. Checking for the absence/presence of "accent" or "warn" is enough, no?

devversion replied:
My thinking is that `class` can contain more classes and that just checking using `includes` is not working 100%. e.g. class attribute could be: `custom-mat-accent mat-primary-field`.

So we need to ensure that there is a space before/after the class we check for, or that it is at the start/end of the `class` attribute value. or just split the `class` attribute by space.. Surely these cases are extremely rare but I don't think there is a benefit in making these things less robust if we can just simply use the public robust `hasClass` methods.

The current implementation basically checks for existence of the theme classes. It just uses the intended APIs but unfortunately looks kind of dirty (to me at least)

mmalerba commented on src/material-experimental/form-field/testing/form-field-harness.spec.ts:
```diff
@@ -0,0 +1,18 @@
+
+// TODO(devversion): we cannot import the select harness through the module name
+// because it does not expose an entry-point that re-exports the harness
+import {MatSelectHarness} from '../../mdc-select/harness/select-harness';
```
This should be in now if you rebase

mmalerba commented on src/material-experimental/form-field/testing/form-field-harness.ts:
```diff
@@ -0,0 +1,231 @@
+import {MatInputHarness} from '@angular/material-experimental/input/testing';
+// TODO(devversion): we cannot import the select harness through the module name
+// because it does not expose an entry-point that re-exports the harness
+import {MatSelectHarness} from '../../mdc-select/harness/select-harness';
```
Can rebase and fix

mmalerba commented on test/karma-system-config.js:
```diff
@@ -116,7 +116,7 @@ System.config({
     '@angular/material/divider': 'dist/packages/material/divider/index.js',
     '@angular/material/expansion': 'dist/packages/material/expansion/index.js',
-    '@angular/material/form-field': 'dist/packages/material/form-field/index.js',
+    '@angular/material/form-field':  'dist/packages/material/form-field/index.js',
```
need to add `/testing` and `/testing/shared.spec`

devversion replied:
done.

mmalerba commented on src/material-experimental/form-field/testing/shared.spec.ts:
```diff
@@ -0,0 +1,262 @@
+
+/** Shared tests to run on both the original and MDC-based form-field's. */
+export function runHarnessTests(
+    modules: Type<any>[], formFieldHarness: typeof MatFormFieldHarness, inputHarness: Type<any>,
```
shouldn't these be `typeof MatInputHarness` and `typeof MatSelectHarness`

devversion replied:
There is no real need to type them as we don't use any methods of those harnesses. We just need to compare the object instances. Importing the types would add an unnecessary dep.

mmalerba commented on src/material-experimental/form-field/testing/form-field-harness.ts:
```diff
@@ -0,0 +1,231 @@
+// TODO(devversion): support datepicker harness once developed (COMP-203).
+// Also support chip list harness.
+/** Possible harnesses of controls which can be bound to a form-field. */
+export type FormFieldControlHarness = MatInputHarness|MatSelectHarness;
```
Hmm, people can also make their own custom form field controls. I wonder if we can handle this somehow...

devversion replied:
Good point. I assume that we can't really detect the harness for custom controls (if there is one), but we can document that for custom form-field controls, the harness needs to be retrieved manually.

Otherwise we could also allow passing in a harness into `getControl(type)` that will return the harness instance for an element that matches the passed harness type. WDYT?

mmalerba replied:
Yeah that might be a good idea, we could make the type optional. If you don't pass anything it can just look for input, select, or datepicker. If you do pass one it will look for just the harness you passed. 

If there's any common methods that the `MatFormFieldHarness` would need to call we can make a `MatFormFieldControlHarness` base class and require people to extend that

devversion replied:
Yes. I like the idea of having a base class for custom form-field control harnesses. I added one that people can implement (even though it is empty for now), but it makes breaking changes less if we ever add stuff there.

devversion commented:
@mmalerba @jelbourn Addressed feedback. please have another look.

mmalerba commented on src/material-experimental/form-field/testing/form-field-harness.ts:
```diff
@@ -0,0 +1,226 @@
+ * Base class for custom form-field control harnesses. Harnesses for
+ * custom controls with form-fields need to implement this interface.
+ */
+export interface CustomFormFieldControlHarness extends ComponentHarness {}
```
Let's call it `MatFormFieldControlHarness` and have `MatInputHarness` and `MatSelectHarness` extend it as well. It might need to go in its own file & build rule to make that work though.

devversion replied:
@mmalerba done. I had to introduce another nested entry-point so that default controls which are also referenced in the `MatFormFieldHarness` can import the base control harness without introducing a circular dependency. For public consumers there is no change. They can still import through `form-field/testing`.

mmalerba commented on src/material-experimental/form-field/testing/form-field-harness.ts:
```diff
@@ -0,0 +1,226 @@
+   * Gets the harness of the control that is bound to the form-field. Searches
+   * for a control that matches the specified harness type.
+   */
+  async getControl<X extends CustomFormFieldControlHarness>(type: ComponentHarnessConstructor<X>):
```
Lets make this able to take a `HarnessPredicate` as well, that way people can do something like `MatInputHarness.with({selector: '#ff-control'})`

devversion replied:
Hmm. I'm not sure about that. My thinking is that if they want to search for specific harnesses anyway, they don't need to go through the harness anyway.

devversion replied:
Done as of Slack discussion.

devversion commented:
@mmalerba @jelbourn Addressed most of the feedback. The discussion for the `getThemeColor` is still open. let me know if I should change it or not. I don't feel too strong about it.
